[FEAT] 멘토용 예약 관리(예약 조회/예약 상태 변경/예약자 전화번호 조회) API 구현

dye0p opened the pull request:
## Issue Number
closed #260

## As-Is
<!-- 문제 상황 정의 -->
- 멘토는 자신이 개설한 멘토링에 대한 예약 내역을 확인할 수 없습니다.
- 멘토는 멘토링 예약에 대한 상태(승인, 거절 등)를 변경할 수 없습니다.
- 멘토는 예약한 멘티의 연락처를 확인할 수 없습니다.

## To-Be
<!-- 변경 사항 -->
- 멘토가 자신이 개설한 모든 멘토링의 예약 목록을 조회하는 기능을 추가했습니다.
- 멘토가 예약 상태를 승인, 거절, 완료 등으로 변경할 수 있는 기능을 추가했습니다.
- 멘토가 예약자의 전화번호를 조회하는 기능을 추가했습니다. (단, 예약이 `PENDING` 상태일 때는 개인정보 보호를 위해 전화번호가 노출되지 않습니다.)
- 위 기능들에 대한 API 엔드포인트를 구현하고 관련 비즈니스 로직 및 테스트 코드를 작성했습니다.


## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
- `Reservation` 엔티티에 `status` 필드를 추가하여 예약 상태를 관리하도록 했습니다.
- `Status` Enum을 통해 예약 상태 변경 로직의 유효성을 검사하도록 구현했습니다.
- `develop` 브랜치와의 충돌을 방지하기 위해 `WebConfig`의 인터셉터 경로에 `/reservations/**`를 추가했습니다.

dye0p committed:
feat: 멘토링의 진행 상태를 표시하기 위해서 상태(status) 필드 추가

dye0p committed:
feat: 상태를 나타내는 Status Enum 객체에 Complete 필드와 value추가

dye0p committed:
refactor: Reservation 객체에서 getStatus() 메서드 반환값을 Status 필드의 value 값으로 변경

dye0p committed:
fix: 생성된 예약의 응답에서 mentorPhone 정보 제거

- 예약이 생성되면 멘토의 전화번호를 노출시키지 않기로 하는 정책상의 이유로 변경

dye0p committed:
fix: 예약 생성시 발송되는 SMS의 내용에서 멘토와 멘티의 전화번호 제거

- 예약이 생성되면 멘토의 전화번호를 노출시키지 않기로 하는 정책상의 이유로 변경

dye0p committed:
feat: 예약이 생성되면, 예약 상태 PENDING을 추가 + 반환값 엔티티 객체로 변경

- 예약생성 응답 dto에서 멘토 전화번호가 제거되었기 때문에 멘토용 SMS에 전송될 전화번호가 없어진 상황. 그래서 ReservationService에서 Reservation 객체를 반환하여 Reservation에서 전화번호를 가져올 수 있도록 변경

dye0p committed:
feat: 멘토가 개설한 멘토링의 모든 예약을 조회하는 기능 구현

- 멘토가 개설한 멘토링의 모든 예약을 조회한다.
- 예약의 상태는 승인대기, 승인, 거부, 완료 상태의 예약이 모두 포함된다.

dye0p committed:
feat: 멘토가 개설한 멘토링의 모든 예약을 조회하는 API 구현

- 조회에 성공하면 200 OK와 함께 리스트를 응답한다.

dye0p committed:
feat: 멘토가 개설한 멘토링의 예약을 승인하는 API 뼈대 코드 추가

- 멘토는 자신이 개설한 멘토링의 예약을 승인할 수 있다.
- 승인을 하게 되면 멘티의 전화번호를 응답한다.

dye0p committed:
refactor: Status enum 필드 상수명 수정(-ed 제거)

dye0p committed:
refactor: Status enum 생성자 필드명 명확하게 변경

dye0p committed:
refactor: REJECT value명 변경(거부 -> 거절)

dye0p committed:
feat: WebConfig 인터셉터 경로 추가(/reservations/**)

dye0p committed:
feat: (멘토링) 예약의 상태를 변경하는 기능 구현

dye0p committed:
feat: 예약자의 전화번호를 요청하는 API 구현

dye0p committed:
Merge branch 'develop' into feature/260-mentor-reservations-manager

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/config/WebConfig.java
#	backend/fittoring/src/main/java/fittoring/exception/GlobalExceptionHandler.java
#	backend/fittoring/src/main/java/fittoring/mentoring/business/exception/BusinessErrorMessage.java
#	backend/fittoring/src/main/java/fittoring/mentoring/business/repository/ReservationRepository.java

dye0p committed:
test: 리뷰 관련 테스트 Reservation 생성자에 누락된 status 추가하여 테스트 수정

dye0p committed:
build: Gradle 테스트 리포트 업로드 코드 추가

dye0p committed:
test: 멘토가 개설한 멘토링의 예약 조회 테스트 createAt 필드 비교 제외

- LocalDateTime 나노초 정밀도 차이로 인한 로컬 테스트와 ci 테스트의 결과가 달라지는 문제를 해결함

kimdevv commented on .github/workflows/backend-ci.yml:
```diff
@@ -74,6 +74,13 @@ jobs:
       - name: Build & Run tests
         run: ./gradlew clean build
 
+      - name: Upload Test Report
+        if: failure()
+        uses: actions/upload-artifact@v4
+        with:
+          name: junit-test-report
+          path: backend/fittoring/build/reports/tests/test/
```
p2: 이게 어떤 기능을 하는지 적어주시면 좋을 것 같습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/exception/GlobalExceptionHandler.java:
```diff
@@ -46,6 +47,16 @@ public ResponseEntity<ErrorResponse> handle(CategoryNotFoundException e) {
     }
 
+    @ExceptionHandler(NotFoundReservationException.class)
+    public ResponseEntity<ErrorResponse> handle(NotFoundReservationException e) {
```
p3: `ReservationNotFoundException`도 있어서 나중에 예외 네이밍도 통일해야 할 것 같네요

dye0p replied:
그렇네요! 중복되는 것 같으니 NotFoundReservationException은 제거하겠습니다.

[refactor: 중복되었던 ReservationNotFoundException 클래스 제거](https://github.com/woowacourse-teams/2025-Fit-toring/pull/316/commits/57fd497785dce21889fcfd135ed003370f81995d)

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java:
```diff
@@ -1,9 +1,34 @@
+    COMPLETE("완료"),
     ;
+
+    private final String value;
+
+    Status(String value) {
+        this.value = value;
+    }
```
p3: `value`는 왜 추가된 건가요?? 상태를 요청받을 때 "APPROVE"와 같이 요청받아도 괜찮지 않나요??

dye0p replied:
맞습니다! 로직상 꼭 필요한 것은 아니지만, 각각의 필드가 어떤 의미인지 명확하게 하기 위해서 value를 추가 하였습니다!

ppparkta replied:
p2: 불필요한 값이라면 지우는게 좋지 않을까요? 한글명은 정해진 컨벤션이 없어서 프론트에서 사용하는 명칭이 달라진다면 오히려 혼란스러워질 수 있을 것 같아요.

jye326 replied:
저도 한글 value 값은 굳이 필요하지 않은 것 같아요!

kimdevv reviewed (approved):
고생하셨습니다~ 간단하게 궁금한 부분들 커멘트 남깁니다!

dye0p committed:
refactor: 중복되었던 ReservationNotFoundException 클래스 제거

dye0p committed:
build: backend-ci 테스트 리포트 업로드 설명 주석 추가

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Reservation.java:
p3 : 이 정도면 이제 Getter를 클래스 레벨에 붙여도 될 것 같네요

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java:
```diff
@@ -1,9 +1,34 @@
+        if (this.equals(updateStatus)) {
+            throw new InvalidStatusException(BusinessErrorMessage.RESERVATION_STATUS_ALREADY_EQUAL.getMessage());
+        }
+    }
```
p2 : Status는 예약 상태 뿐만 아니라 자격증, 멘토링 등의 승인, 대기, 거절 상태에도 사용됩니다.
이 validate 메서드는 예약 상태 검증을 하는 것 같은데 Status의 범용성을 위해 validateReservationStatus 등으로 구체화 해야 할 것 같습니다!

dye0p committed:
fix: 멘토에게 보낼 sms 메세지 내용을 변경

- 멘토에게 보낼 문자내용에 멘토의 이름이 들어가 있던 문제 수정

dye0p committed:
refactor: SmsReservationMessage 클래스 제거

dye0p committed:
fix: 예약 완료시 응답 수정

- 상담 내용, 상태 제거

dye0p committed:
refactor: 멘토링 예약 로직 리팩터링

dye0p committed:
refactor:  Reservation 컬럼명 명확하게 변경(context -> content)

dye0p committed:
refactor: data.sql 네이밍 변경

dye0p committed:
fix: ci 코드 주석 제거

dye0p committed:
fix:  ci test

ppparkta commented on backend/fittoring/src/main/java/fittoring/exception/GlobalExceptionHandler.java:
p1: NotFoundReservationException은 핸들링되고 있지 않네요

dye0p replied:
기존에 `ReservationNotFoundException` 커스텀 예외가 존재해서`NotFoundReservationException`은  제거하였습니다!

[refactor: 중복되었던 ReservationNotFoundException 클래스 제거](https://github.com/woowacourse-teams/2025-Fit-toring/pull/316/commits/57fd497785dce21889fcfd135ed003370f81995d)

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Reservation.java:
```diff
@@ -30,12 +32,14 @@ public class Reservation {
     private LocalDateTime createdAt;
 
+    @Getter
     @ManyToOne
     @JoinColumn(nullable = false)
     private Mentoring mentoring;
```
p3: 연관관계 매핑은 하단으로 내려서 모아보기 편하게 하는거 어떠신가요?

dye0p replied:
좋아요! 그런데 코드에서 궁금한 부분이 있는데, Reservation의 생성자에서 createdAt을 null로 받고 있는데, 의도가 궁금합니다!

``` java
    public Reservation(String content, Mentoring mentoring, Member mentee, Status status) {
        this(null, content, null, mentoring, mentee, status);
    }
```

[refactor: Reservation 엔티티의 필드 순서 정리](https://github.com/woowacourse-teams/2025-Fit-toring/pull/316/commits/2bd514f3be018f7feaa63b5e995e7a58a127729b)

ppparkta replied:
영속화되는 시점에 이벤트를 감지해서 자동 주입하기 때문에 비워뒀습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java:
```diff
@@ -1,9 +1,34 @@
+    APPROVE("승인"),
+    PENDING("대기"),
+    REJECT("거절"),
+    COMPLETE("완료"),
```
p3: 각 enum의 값도 기존값으로 변경되는 것이 올바른 것 같아요! -ed 붙이는거요!

dye0p replied:
수정했습니다! 영어에 약해서 죄송합니다..

[refactor: Status Enum 객체 상수명에 -ed 추가](https://github.com/woowacourse-teams/2025-Fit-toring/pull/316/commits/3de8c8951080627b5c3fee8622899e7c49adf152)

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/ReservationService.java:
```diff
@@ -24,18 +29,74 @@ public class ReservationService {
-        return ReservationCreateResponse.from(savedReservation);
+    }
+
+    @Transactional(readOnly = true)
```
p3: readOnly 옵션을 사용하신 이유가 있나요?

dye0p replied:
메서드 안에서 조회를 두번 수행하는데, 중간에 예상치 못한 변경이 일어나도 db의 변경이 일어나지 않게 하기 위해서 readOnly옵션을 true로 지정했습니다!

ppparkta reviewed (commented):
고생하셨어요~~! 
기존에 Status enum의 값을 처리하고 있었던 프론트 코드(관리자페이지)가 있기 때문에 값을 되돌려주세요. 그리고 이런 값 수정이 불가피하게 발생하면 PR본문에 남겨주시면 감사하겠습니다!

dye0p committed:
Merge branch 'develop' into feature/260-mentor-reservations-manager

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/mentoring/business/model/Reservation.java
#	backend/fittoring/src/main/java/fittoring/mentoring/business/repository/ReservationRepository.java
#	backend/fittoring/src/main/java/fittoring/mentoring/business/service/ReservationService.java
#	backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/ReservationController.java
#	backend/fittoring/src/test/java/fittoring/integration/mentoring/api/ReservationControllerTest.java
#	backend/fittoring/src/test/java/fittoring/mentoring/business/service/ReservationServiceTest.java

dye0p committed:
feat: 인터셉터 필터링 경로 추가

dye0p committed:
refactor: Status Enum 객체 상수명에 -ed 추가

dye0p committed:
test: 테스트 픽스처 수정

dye0p committed:
refactor: import문 변경

dye0p committed:
Merge branch 'develop' into feature/260-mentor-reservations-manager

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/mentoring/business/exception/BusinessErrorMessage.java

dye0p committed:
test: 멘토가 개설한 멘토링에 대한 예약을 조회하는 테스트에서 createdAt 필드 검사 제외

- createdAt의 타입은 LocalDateTime으로, 나노초 까지 기록한다.
- 로컬 테스트 결과와, CI 테스트의 결과가 나노초 차이로 달라지는 문제가 있었음
- 통합 테스트에서 나노초 차이까지 검증이 필요하지 않다고 판단하였음

dye0p committed:
refactor: Reservation 엔티티의 필드 순서 정리

dye0p committed:
refactor: Status Enum 상수 필드명 제거

jye326 commented on backend/fittoring/src/main/java/fittoring/config/auth/AuthenticationInterceptor.java:
```diff
@@ -22,12 +22,15 @@ public class AuthenticationInterceptor implements HandlerInterceptor {
             throws Exception {
 
-        if(request.getMethod().equalsIgnoreCase("OPTIONS")){
+        if (request.getMethod().equalsIgnoreCase("OPTIONS")) {
```
👍

jye326 commented on backend/fittoring/src/main/java/fittoring/config/auth/AuthenticationInterceptor.java:
```diff
@@ -22,12 +22,15 @@ public class AuthenticationInterceptor implements HandlerInterceptor {
 
-        if (request.getMethod().equalsIgnoreCase("GET") &&
-                request.getRequestURL().toString().contains("/mentorings")) {
+        if (
+                request.getMethod().equalsIgnoreCase("GET") &&
+                request.getRequestURL().toString().contains("/mentorings") &&
+                !request.getRequestURL().toString().contains("/mine")
+        ) {
```
p3: 이렇게 보니까 인증 부분 리팩터링이 시급해보이네요... 회의 때 이슈로 추가해보아요~

jye326 reviewed (commented):
리뷰 확인이 늦어져서 죄송합니다...! 코멘트 남겼으니 확인 부탁드려요!

ppparkta reviewed (approved):
고생하셨습니다!

dye0p committed:
refactor: 예약 상태 검증 메서드 네이밍 변경

jye326 reviewed (approved):
승인합니다~
